soroban-sdk/src/deploy.rs

```diff
@@ -53,61 +56,102 @@ impl Deployer {
-    /// given contract ID and the provided salt.
-    pub fn with_other_contract(
+    /// Get a deployer that deploys contracts that derive the contract ID
+    /// from the provided address and salt.
+    ///
+    /// The deployer address must authorize all the deployments.
+    pub fn with_address(
+        &self,
+        address: Address,
+        salt: impl IntoVal<Env, BytesN<32>>,
+    ) -> DeployerWithAddress {
+        DeployerWithAddress {
+            env: self.env.clone(),
+            address,
+            salt: salt.into_val(&self.env),
+        }
+    }
+
+    /// Get a deployer that deploys an instance of Stellar Asset Contract
+    /// corresponding to the provided serialized asset.
+    ///
+    /// `serialized_asset` is the Stellar `Asset` XDR serialized to bytes. Refer
+    /// to `stellar-xdr` create for the exact definition.
+    pub fn with_stellar_asset(
         &self,
-        contract_id: &Address,
-        salt: &impl IntoVal<Env, BytesN<32>>,
-    ) -> DeployerWithOtherContract {
-        let env = self.env();
-        DeployerWithOtherContract {
-            env: env.clone(),
-            contract_id: contract_id.clone(),
-            salt: salt.into_val(env),
+        serialized_asset: impl IntoVal<Env, Bytes>,
+    ) -> DeployerWithAsset {
+        DeployerWithAsset {
+            env: self.env.clone(),
+            serialized_asset: serialized_asset.into_val(&self.env),
         }
     }
+
+    /// Upload the contract Wasm code to the network.
+    ///
+    /// Returns the hash of the uploaded Wasm that can be then used for
+    /// the contract deployment.
+    /// ### Examples
+    /// ```
+    /// use soroban_sdk::{BytesN, Env};
+    ///
+    /// const WASM: &[u8] = include_bytes!("../doctest_fixtures/contract.wasm");
+    ///
+    /// #[test]
+    /// fn test() {
+    /// # }
+    /// # fn main() {
+    ///     let env = Env::default();
+    ///     env.deployer().upload_contract_wasm(WASM);
+    /// }
+    /// ```
+    pub fn upload_contract_wasm(&self, contract_wasm: impl IntoVal<Env, Bytes>) -> BytesN<32> {
+        self.env
+            .upload_wasm(contract_wasm.into_val(&self.env).to_object())
+            .unwrap_infallible()
+            .into_val(&self.env)
+    }
 }
 
-/// A deployer that deploys a contract that has its ID derived from the current
-/// contract ID and the provided salt.
-pub struct DeployerWithCurrentContract {
+/// A deployer that deploys a contract that has its ID derived from the provided
+/// address and salt.
+pub struct DeployerWithAddress {
     env: Env,
+    address: Address,
     salt: BytesN<32>,
 }
 
-impl DeployerWithCurrentContract {
-    /// Return the ID of the contract defined by the deployer.
+impl DeployerWithAddress {
+    /// Return the address of the contract defined by the deployer.
     #[doc(hidden)]
-    pub fn id(&self) -> Address {
+    pub fn contract_address(&self) -> Address {
```

Comment:
Probably sufficient? +1 for Address over ID.
```suggestion
    pub fn address(&self) -> Address {
```
I guess both contract_address and address are pretty ambiguous given we have the address of the deployer 😕.

Maybe:


```suggestion
    /// Returns what the address of the contract will be once deployed.
    pub fn deployed_address(&self) -> Address {
```